pkg/kv/kvserver/replica_circuit_breaker.go

```diff
@@ -0,0 +1,183 @@
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/circuit"
+	"github.com/cockroachdb/cockroach/pkg/util/contextutil"
+	"github.com/cockroachdb/cockroach/pkg/util/envutil"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/errors"
+	"github.com/cockroachdb/redact"
+)
+
+type replicaInCircuitBreaker interface {
+	Clock() *hlc.Clock
+	Desc() *roachpb.RangeDescriptor
+	Send(context.Context, roachpb.BatchRequest) (*roachpb.BatchResponse, *roachpb.Error)
+	slowReplicationThreshold(ba *roachpb.BatchRequest) (time.Duration, bool)
+	rangeUnavailableError() error
+}
+
+var defaultReplicaCircuitBreakerSlowReplicationThreshold = envutil.EnvOrDefaultDuration(
+	"COCKROACH_REPLICA_CIRCUIT_BREAKER_SLOW_REPLICATION_THRESHOLD", 0,
+)
+
+var replicaCircuitBreakerSlowReplicationThreshold = settings.RegisterPublicDurationSettingWithExplicitUnit(
+	settings.SystemOnly,
+	"kv.replica_circuit_breaker.slow_replication_threshold",
+	"duration after which slow proposals trip the per-Replica circuit breaker (zero duration disables breakers)",
+	defaultReplicaCircuitBreakerSlowReplicationThreshold,
+	func(d time.Duration) error {
+		if d < 0 {
+			return errors.New("duration must be non-negative")
+		}
+		return nil
+	},
+)
+
+type replicaCircuitBreaker struct {
+	ambCtx  log.AmbientContext
+	stopper *stop.Stopper
+	r       replicaInCircuitBreaker
+	st      *cluster.Settings
+	wrapped *circuit.Breaker
+}
+
+func (br *replicaCircuitBreaker) enabled() bool {
+	return replicaCircuitBreakerSlowReplicationThreshold.Get(&br.st.SV) > 0 &&
+		br.st.Version.IsActive(context.Background(), clusterversion.ProbeRequest)
+}
+
+func (br *replicaCircuitBreaker) newError() error {
+	return br.r.rangeUnavailableError()
+}
+
+func (br *replicaCircuitBreaker) TripAsync() {
+	if !br.enabled() {
+		return
+	}
+
+	_ = br.stopper.RunAsyncTask(
+		br.ambCtx.AnnotateCtx(context.Background()), "trip-breaker",
+		func(ctx context.Context) {
+			br.wrapped.Report(br.newError())
+		},
+	)
+}
+
+type neverTripSignaller struct{}
+
+func (s neverTripSignaller) Err() error         { return nil }
+func (s neverTripSignaller) C() <-chan struct{} { return nil }
+
+func (br *replicaCircuitBreaker) Signal() interface {
+	Err() error
+	C() <-chan struct{}
+} {
+	if !br.enabled() {
+		return neverTripSignaller{}
+	}
+	return br.wrapped.Signal()
+}
+
+func newReplicaCircuitBreaker(
+	cs *cluster.Settings,
+	stopper *stop.Stopper,
+	ambientCtx log.AmbientContext,
+	r replicaInCircuitBreaker,
+) *replicaCircuitBreaker {
+	br := &replicaCircuitBreaker{
+		stopper: stopper,
+		ambCtx:  ambientCtx,
+		r:       r,
+		st:      cs,
+	}
+
+	br.wrapped = circuit.NewBreaker(circuit.Options{
+		Name:       redact.Sprintf("breaker"), // log bridge has ctx tags
```

Comment:
nit: unnecessary `Sprintf()`, a literal string will autocast to `redact.RedactableString`.